### PR TITLE
Close idle HTTP connections after 3 minutes

### LINKED ISSUE #14951: Timeouts when using the remote Build Cache in a GitHub workflow
We're running builds via GitHub Actions using a remote http build cache (on an AWS EC2 instance). For long running builds with many tests, the HttpBuildCacheService runs into connection issues, mostly `Connection timed out (Write failed)` and sometimes `BuildCacheException: Read timed out`. Running from a local machine or from an internal TeamCity agent works without issues. Also, faster builds (faster/less tests) can successfully update the remote build cache.

### Expected Behavior

We'd like to have less timeouts or automated retries when Gradle tries to update the build cache.

### Current Behavior

The build doesn't fail, but doesn't update the remote build cache. So, subsequent builds cannot reuse cache entries.

### Context

We tried to work around the issue by configuring the HttpClient. Our most recent config looked like this:

command line:
```
./gradlew clean build --no-watch-fs --info --debug --stacktrace -Dhttp.keepAlive=false -Dorg.gradle.internal.http.connectionTimeout=60000 -Dorg.gradle.internal.http.socketTimeout=60000
```

gradle.properties:
```
org.gradle.caching=true
```

settings.gradle:
```
buildCache {
  val isCiServer = System.getenv().containsKey("CI")
  local {
    isEnabled = !isCiServer
  }
  remote<HttpBuildCache> {
    url = uri("${System.getenv()["GRADLE_BUILD_CACHE_URI"]}/cache/")
    isEnabled = true
    isPush = isCiServer

    credentials {
        username = System.getenv()["GRADLE_BUILD_CACHE_USERNAME"]
        password = System.getenv()["GRADLE_BUILD_CACHE_PASSWORD"]
    }
  }
}
```

### Steps to Reproduce 

... we'll try to provide one.

### Your Environment

Build scan URL:
We don't use build scans.

Debug log (partial):

```
... still cleaning up, I'll update the description later with the debug log.
```

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Automatically close idle connections after 3 min

Closes #14951
Related to #14943

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
4 minutes 33 seconds would be more fitting for the timeout.

**Comment by JLLeitschuh** [comment]
I'm trying to do something under the Azure default of 4 min. Luke proposed 3 min.

https://gradle.slack.com/archives/CHDLT99C6/p1638440509132600?thread_ts=1636477584.059200&channel=CHDLT99C6&message_ts=1638440509.132600

**Comment by JLLeitschuh** [comment]
Reduced timeout to 3 min, as suggested by @ldaley

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

File: subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettingsTest.groovy [hunk 1]
```diff
@@ -65,4 +75,15 @@ class JavaSystemPropertiesHttpTimeoutSettingsTest extends Specification {
...
+
+    def "uses default value if provided idle connection timeout is not valid"() {
+        System.setProperty(IDLE_CONNECTION_TIMEOUT_SYSTEM_PROPERTY, timeout)
+        JavaSystemPropertiesHttpTimeoutSettings settings = new JavaSystemPropertiesHttpTimeoutSettings()
+
+        expect:
+        settings.idleConnectionTimeoutMs == DEFAULT_IDLE_CONNECTION_TIMEOUT
+
+        where:
+        timeout << ["", "abc"]
```
Comment (tresat) [inline subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettingsTest.groovy:87]: There's no invalid number warning to check for here?

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettings.java [hunk 2]
```diff
@@ -20,19 +20,31 @@
...
     public static final String SOCKET_TIMEOUT_SYSTEM_PROPERTY = "org.gradle.internal.http.socketTimeout";
+    public static final String IDLE_CONNECTION_TIMEOUT_SYSTEM_PROPERTY = "org.gradle.internal.http.idleConnectionTimeout";
     public static final int DEFAULT_CONNECTION_TIMEOUT = 30000;
     public static final int DEFAULT_SOCKET_TIMEOUT = 30000;
+    /**
+     * The default time in milliseconds for an idle connection to remain open.
+     * <a href="https://azure.microsoft.com/en-us/blog/new-configurable-idle-timeout-for-azure-load-balancer/">Microsoft Azure closes idle connections after 4 min</a>,
+     * so we set our default to be below that.
+     */
+    public static final int DEFAULT_IDLE_CONNECTION_TIMEOUT = (int) Duration.ofMinutes(3).toMillis();
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettings.java:38]: This makes sense if we find this is inadequate. But for now, we've only seen this issue reported for Azure.

File: subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettingsTest.groovy (hunk 1 above)
Comment (JLLeitschuh) [reply subprojects/resources-http/src/test/groovy/org/gradle/internal/resource/transport/http/JavaSystemPropertiesHttpTimeoutSettingsTest.groovy:87]: No, we don't have one here or for any of the other values either. I think one reason for that is that these are supposed to be "internal" flags.

**Comment by JLLeitschuh** [comment]
@bot-gradle test and merge

**Comment by jaswanthikolla** [comment]
This is a wrong fix. 

As per documentation of [setConnectionTimeToLive](https://javadoc.io/doc/org.apache.httpcomponents/httpclient/4.5.2/org/apache/http/impl/client/HttpClientBuilder.html#setConnectionTimeToLive(long,%20java.util.concurrent.TimeUnit)) ,  It defines the maximum time for a connection to live/used for, it's got nothing to do Idle connection.

To close Idle HTTP connections, we need to use [setKeepAliveStrategy](https://javadoc.io/doc/org.apache.httpcomponents/httpclient/4.5.2/org/apache/http/impl/client/HttpClientBuilder.html#setKeepAliveStrategy(org.apache.http.conn.ConnectionKeepAliveStrategy)) , and use something like following.

```
ConnectionKeepAliveStrategy keepAliveStrategy = new ConnectionKeepAliveStrategy() {
            @Override
            public TimeValue getKeepAliveDuration(HttpResponse response, HttpContext context) {
                // Set keep-alive timeout to 100 seconds
                return TimeValue.ofSeconds(100);
            }
        };
```

**Comment by jaswanthikolla** [comment]
This is a wrong fix. 

As per documentation of [setConnectionTimeToLive](https://javadoc.io/doc/org.apache.httpcomponents/httpclient/4.5.2/org/apache/http/impl/client/HttpClientBuilder.html#setConnectionTimeToLive(long,%20java.util.concurrent.TimeUnit)) ,  It defines the maximum time for a connection to live/used for, it's got nothing to do Idle connection.

To close Idle HTTP connections, we need to use [setKeepAliveStrategy](https://javadoc.io/doc/org.apache.httpcomponents/httpclient/4.5.2/org/apache/http/impl/client/HttpClientBuilder.html#setKeepAliveStrategy(org.apache.http.conn.ConnectionKeepAliveStrategy)) , and use something like following.

```
ConnectionKeepAliveStrategy keepAliveStrategy = new ConnectionKeepAliveStrategy() {
            @Override
            public TimeValue getKeepAliveDuration(HttpResponse response, HttpContext context) {
                return TimeValue.ofSeconds(3 * 60);
            }
        };
```

**Comment by tresat** [comment]
Hi @jaswanthikolla!  Thanks for bringing this to our attention.  Can you please open a new issue for this so we can assess it per our triaging process?

FTR: new issues are almost always a better way to share concerns with us versus new comments on closed issues, which might not be seen.

**Pull request merged**